adapter/src/main/java/com/twilio/kudu/dataloader/DataLoader.java

```diff
@@ -263,67 +276,95 @@ private void bindValues(PreparedStatement statement, UniformLongValueGenerator t
+    long threadDelta = Math.max(rangePerThread, cubeGranularityFloorMod);
     List<CalciteKuduTable> cubeTables = calciteKuduTable.getCubeTables();
     int numRows = numRowsOverrideOption.orElseGet(scenario::getNumRows);
     for (int t = 0; t < threadPoolSize; ++t) {
-      long floorMod = cubeTables.isEmpty() ? DateTimeUtils.MILLIS_PER_SECOND : cubeGranularityFloorMod;
       // for each thread the start time stamp is the previous threads end timestamp
-      // (other than
-      // the first thread for which we set it to the scenario min timestamp)
-      startTimestamp = endTimestamp;
+      // (except the first thread for which we set it to the scenario min timestamp)
       // for each thread the end timestamp is set to the floor of the (start timestamp
-      // + delta) to
-      // DAY so that cubes are aggregated correctly (other than the last thread for
-      // which we set
-      // it to the scenario max timestamp)
+      // + delta)
+      // truncated to the max cube granularity DAY so that cubes are aggregated
+      // correctly
+      // (except the last thread for which we set it to the scenario max timestamp)
       boolean lastThread = t == threadPoolSize - 1;
-      endTimestamp = lastThread ? scenarioEndTimestamp : SqlFunctions.floor(startTimestamp + delta, floorMod);
+      final long threadStartTimestamp = prevThreadEndTimestamp;
+      final long threadEndTimestamp = lastThread ? scenarioEndTimestamp
+          : SqlFunctions.floor(threadStartTimestamp + threadDelta, cubeGranularityFloorMod);
+
+      // further divide the time range that each thread is responsible for so that we
+      // limit
+      // the amount of state in CubeMutationState
+      int numTasksPerThread = (int) Math
+          .ceil((float) (threadEndTimestamp - threadStartTimestamp) / cubeGranularityFloorMod);
 
-      long finalStartTimestamp = startTimestamp;
-      long finalEndTimestamp = endTimestamp;
       // callable implemented as lambda expression
+      final int threadIndex = t;
       Callable<Void> callableObj = () -> {
-        logger.info("startTimestamp {}", new Date(finalStartTimestamp));
-        logger.info("endTimestamp {}", new Date(finalEndTimestamp));
-        int numRowsWrittenPerThread = numRows / threadPoolSize;
+        logger.info("thread{} start timestamp {}", threadIndex, new Date(threadStartTimestamp));
+        logger.info("thread{} end timestamp {}", threadIndex, new Date(threadEndTimestamp));
+        int numRowsPerThread = numRows / threadPoolSize;
         if (lastThread) {
           // let the last thread write any remaining rows
-          numRowsWrittenPerThread += numRows % threadPoolSize;
+          numRowsPerThread += numRows % threadPoolSize;
         }
+        int maxBatchesPerThread = (int) Math.ceil((float) (numRowsPerThread) / CUBE_MUTATION_SIZE);
+        int numCubeMutationBatches = Math.min(numTasksPerThread, maxBatchesPerThread);
```

Comment:
Partitions the data to be written by each thread into batches. The number of rows written in each batch will be either CUBE_MUTATION_SIZE if we have enough days/hours per thread.